discovery_service panics when EKS matcher uses assume_role_arn

I have discovery service running with an `eks` matcher type that uses `assume_role_arn`, e.g.
```yaml
discovery_service:
  enabled: true
  discovery_group: "teleport-dev-2"
  aws:
    - types:
        - "eks"
      regions:
        - "ca-central-1"
      assume_role_arn: "arn:aws:iam::123123123123:role/gavin-alpha-devteleport-com-oidc"
```

- If my EKS cluster does not have an access entry for that role, then Teleport panics.
- If it already has an access entry, it does not panic.
- If I do not use `assume_role_arn` and instead assume the role with env vars manually (teleport is running on my local machine), then it also does not panic.

Another detail that might be necessary to reach this bug is that I updated my IAM role's permissions to include `eks:ListAccessEntries`, which is a permission we use but do not document (currently).

## details

- Teleport version: 16.1.1
- Debug logs:
```
teleport start -c /Users/gavin/code/dev/clusters/alpha/agent.yaml -d
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x98 pc=0x109949694]

goroutine 929 [running]:
github.com/gravitational/teleport/lib/kube/utils.GenAWSEKSToken({0x0?, 0x0?}, {0x140021c2c50, 0xc}, {0x10f0f2c50, 0x1146b0c80})
	github.com/gravitational/teleport/lib/kube/utils/eks_token_signed.go:45 +0x34
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).createKubeClient(0x14000c00008?, 0x140018fb500)
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:566 +0x60
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).upsertRoleAndBinding(0x140017cac30, {0x10f0d9040, 0x14001f6caf0}, 0x140018fb500)
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:548 +0x40
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).temporarilyGainAdminAccessAndCreateRole(0x140017cac30, {0x10f0d9040, 0x14001f6caf0}, {0x10f18efe8, 0x140000f4e00}, 0x140018fb500)
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:531 +0x5ac
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).checkOrSetupAccessForARN(0x140017cac30, {0x10f0d9040, 0x14001f6caf0}, {0x10f18efe8, 0x140000f4e00}, 0x140018fb500)
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:430 +0x344
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).getMatchingKubeCluster(0x140017cac30, {0x10f0d9040, 0x14001f6caf0}, {0x140022d6450, 0xc})
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:376 +0x3c4
github.com/gravitational/teleport/lib/srv/discovery/fetchers.(*eksFetcher).getEKSClusters.func1.1()
	github.com/gravitational/teleport/lib/srv/discovery/fetchers/eks.go:284 +0x60
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 618
	golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0x98
```

cc @tigrato

--- a/lib/srv/discovery/fetchers/eks.go
+++ b/lib/srv/discovery/fetchers/eks.go
@@ -563,6 +563,9 @@ func (a *eksFetcher) upsertRoleAndBinding(ctx context.Context, cluster *eks.Clus
 }
 
 func (a *eksFetcher) createKubeClient(cluster *eks.Cluster) (*kubernetes.Clientset, error) {
+	if a.stsClient == nil {
+		return nil, trace.BadParameter("STS client is not set")
+	}
 	token, _, err := kubeutils.GenAWSEKSToken(a.stsClient, aws.StringValue(cluster.Name), a.Clock)
 	if err != nil {
 		return nil, trace.Wrap(err, "unable to generate EKS token for cluster %q", aws.StringValue(cluster.Name))
@@ -666,10 +669,6 @@ func (a *eksFetcher) upsertAccessEntry(ctx context.Context, client eksiface.EKSA
 }
 
 func (a *eksFetcher) setCallerIdentity(ctx context.Context) error {
-	if a.AssumeRole.RoleARN != "" {
-		a.callerIdentity = a.AssumeRole.RoleARN
-		return nil
-	}
 	var err error
 	a.stsClient, err = a.ClientGetter.GetAWSSTSClient(
 		ctx,
@@ -678,6 +677,11 @@ func (a *eksFetcher) setCallerIdentity(ctx context.Context) error {
 	)
 	if err != nil {
 		return trace.Wrap(err)
+	}
+
+	if a.AssumeRole.RoleARN != "" {
+		a.callerIdentity = a.AssumeRole.RoleARN
+		return nil
 	}
 	identity, err := a.stsClient.GetCallerIdentityWithContext(ctx, &sts.GetCallerIdentityInput{})
 	if err != nil {

